### DISCUSSION TITLE
2FA methods

### DISCUSSION THREAD

**Discussion by scottypitcher** [community/community#65247] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

Hi,
I'm happy to use 2FA for account access, but I've stopped using mobile phone completely the last couple of years. I've got a land line at both desks I sit at, and our own email server here for business emails. I'm happy to use those methods but I've deliberately opted out of mobile phone and won't go back.
Kind regards,
Scott

**Comment by queenofcorgis** [comment]
Hi Scott! 

Thanks for asking! Phone-free options include a desktop TOTP app or password manager including TOTP (like 1password) and security keys (fingerprint reader, WebAuthn compatible security key) so you can keep your account safe no phone needed.

**Reply by scottypitcher** [reply]
Thanks for that information. Can you point me to an Open Source or reference implementation of TOTP that I can compile and test?

**Reply by queenofcorgis** [reply]
You're welcome! Our [documentation](https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/configuring-two-factor-authentication#configuring-two-factor-authentication-using-a-totp-mobile-app) goes through how to set up and to refine your search to meet your needs. I don't have a specific one I could recommend, but I saw @ldez  recommend [KeePassXC](https://keepassxc.org/).

**Reply by ldez** [reply]
Hello,

yes you can use [KeypassXC](https://keepassxc.org/) which is free and open source and works offline.

Any [RFC6238](https://datatracker.ietf.org/doc/html/rfc6238) compliant implementation will work, you can find several implementation depending of your language on GitHub or creating your own version.

I also recommend secured keys like [Yubikey](https://www.yubico.com), you can use it as OTP storage and/or hardware security key.

Note: FYI I'm not a GitHub staff member, I'm just a GitHub user.

**Reply by ldez** [reply]
Hello @queenofcorgis, there is a typo inside this [post](https://github.com/orgs/community/discussions/68240):

> by community member @ldezas, which is an open,

FYI, there is a browser integration for KeepassXC (made by the KeepassXC team):
- https://addons.mozilla.org/fr/firefox/addon/keepassxc-browser/
- https://chrome.google.com/webstore/detail/keepassxc-browser/oboonakemofpalcgghocfoadofidjkkk

**Reply by queenofcorgis** [reply]
Thank you for the extra set of eyes and for more helpful links - you're a rockstar 🌟

**Reply by ghost** [reply]
Even better, instead of putting up with Github's garbage 2FA policy simply move to a vendor that is sane like Gitlab.
Don't forget to delete all your repos so they can't train their co-pilot AI on your work anymore than they probably have.

**Comment by tradesouthwest** [comment]
Not sure even where to start on how messed up this 2FA is from Github. I too have no scanner to phone app and legally refuse to connect my phone to ANY service. It is now either move to GitLab or Bitbucket or figure out what people are doing that can not get this to work.

Have tried:
- QR scanner for webcam or screenshot
- Extension to read data from qr code
- Pasting link of qr code result into address bar

So far none of the several Extension I tried worked. 7 or 8, by last count. They all pick up the data and when I try to execute the data (like copy link or copy JUST the "Setup Key" I get a non responsive form submit and can not submit form.

If anyone knows how to scan a QR code from within the browser (other than an extension that does not work) please let me know.

**Reply by queenofcorgis** [reply]
I'm sorry to hear you're having trouble I'd be frustrated too! We do have instructions to configure TOTP manually:

If you wish to setup a TOTP app manually, and require the parameters encoded in the QR code, they are:

Type: `TOTP`
Label: `GitHub:<username>` where `<username>` is your handle on GitHub, for example `monalisa`
Secret: This is the encoded setup key, shown if you click"setup key" during configuration
Issuer: `GitHub`
Algorithm: The default of SHA1 is used
Digits: The default of 6 is used
Period: The default of 30 (seconds) is used

here is the [documentation](https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/configuring-two-factor-authentication) for further assistance. 

It may also help to let us know which app you are using.

**Reply by tradesouthwest** [reply]
Thanks for your response, @queenofcorgis - I did get this to work. Although I am still not a fan of this inconvenience as it is now a requirement. I am guessing this is due to either legal "protection" on MS's side or just another attempt at actually being secure. Either way it MUST be an option and not a requirement.

**Comment by supervisitor** [comment]
Shouldn't one have first thought about the sense and benefit of 2FA... instead of greeting the decision of GitHub/MS immediately with clapping hands, compiling a password manager and time expenditure?
E.g. for me, I see neither need nor sense in the dictate. I think 2FA as an autonomous/on demand activatable option would have made much more sense.

**Comment by tomek-o** [comment]
It looks like TOTP is not complicated, it would be essentially a single OpenSSL function call or a few hundred C code lines (mostly SHA-1). I've made today small Win32/GUI generator, so far tested on my account: https://github.com/tomek-o/totp-generator. In my opinion 2FA it is really pointless though if not used from another/separate device - it is just another password, arbitrary assigned by the service, impossible to remember so it would need to stored somewhere, very likely in the same way as the first/main password.

**Comment by AwokeKnowing** [comment]
I recommend everyone leave github

They are now just become bullies and another Microsoft app. 

They are unfit to host the open source community's work. Now they don't even let you decide how to authenticate. they force you to tie your account to SMS or some app. 

Move on.  

Github has lost it's way.Forcing 2fa against dev's will is EXACTLY the kind of behavior that made MS infamous. Why kill github? why don't you get it? MAKE 2FA OPTIONAL
